### PR TITLE
Fix gather_op to handle various input data types

### PR DESCRIPTION
**Pull request by rjknight**
-GPU based gather operations return incorrect results if
 the input data type is not int32t. Attempts to use other
 data types in the input tensor results in incorrect or
 inconsistent results. This commit modifies the gather
 operation to handle the following data types:
    - int8_t
    - int16_t
    - int32_t
    - int64_t
    - long
    - float
    - double

### CODE REVIEW

**Comment by ezyang** [comment]
Tests failures are real:

```

Jan 16 16:48:47 
Jan 16 16:48:47 =================================== FAILURES ===================================
Jan 16 16:48:47 __________________ TestQuantize8bits.test_SparseSegmentUint8 ___________________
Jan 16 16:48:47 
Jan 16 16:48:47 self = <caffe2.python.lengths_reducer_rowwise_8bit_ops_test.TestQuantize8bits testMethod=test_SparseSegmentUint8>
Jan 16 16:48:47 
Jan 16 16:48:47     def test_SparseSegmentUint8(self):
Jan 16 16:48:47     
Jan 16 16:48:47         init_net = core.Net("init")
Jan 16 16:48:47         net = core.Net("bench")
Jan 16 16:48:47         size = 10**3
Jan 16 16:48:47         isize = 10**2
Jan 16 16:48:47     
Jan 16 16:48:47         # input preparation
Jan 16 16:48:47         d = init_net.UniformFill([], shape=[size, 32])
Jan 16 16:48:47         w = init_net.UniformFill([], shape=[isize, ])
Jan 16 16:48:47         i = init_net.UniformIntFill([], shape=[isize], max=size - 1)
Jan 16 16:48:47         i = init_net.Cast([i], to=core.DataType.INT64)
Jan 16 16:48:47         l = init_net.ConstantFill(
Jan 16 16:48:47             [],
Jan 16 16:48:47             ['l'],
Jan 16 16:48:47             shape=[isize // 10],
Jan 16 16:48:47             value=10,
Jan 16 16:48:47             dtype=core.DataType.INT32,
Jan 16 16:48:47         )
Jan 16 16:48:47         net.FloatToRowwiseQuantized8Bits([d],
Jan 16 16:48:47                                          ['quantized_data', 'scale_bias'])
Jan 16 16:48:47         net.Rowwise8BitQuantizedToFloat(['quantized_data', 'scale_bias'],
Jan 16 16:48:47                                         ['dequantized_data'])
Jan 16 16:48:47     
Jan 16 16:48:47         # SparseLengthsWeightedSum
Jan 16 16:48:47         net.SparseLengthsWeightedSum(['dequantized_data', w, i, l],
Jan 16 16:48:47                                      ['PositionWeighted_0'], engine='fp16')
Jan 16 16:48:47         net.SparseLengthsWeightedSum8BitsRowwise(
Jan 16 16:48:47             ['quantized_data', w, i, l, 'scale_bias'],
Jan 16 16:48:47             ['PositionWeighted_1'])
Jan 16 16:48:47     
Jan 16 16:48:47         # SparseLengthsSum
Jan 16 16:48:47         net.SparseLengthsSum(['dequantized_data', i, l],
Jan 16 16:48:47                              ['Sum_0'], engine='fp16')
Jan 16 16:48:47     
Jan 16 16:48:47         net.SparseLengthsSum8BitsRowwise(
Jan 16 16:48:47             ['quantized_data', i, l, 'scale_bias'],
Jan 16 16:48:47             ['Sum_1'])
Jan 16 16:48:47     
Jan 16 16:48:47         # SparseLengthsWeightedMean
Jan 16 16:48:47         # net.SparseLengthsWeightedMean(['dequantized_data', w, i, l],
Jan 16 16:48:47         #                              ['WeightedMean_0'])
Jan 16 16:48:47         # net.SparseLengthsWeightedMean8BitsRowwise(
Jan 16 16:48:47         #     ['quantized_data', w, i, l, 'scale_bias'],
Jan 16 16:48:47         #     ['WeightedMean_1'])
Jan 16 16:48:47     
Jan 16 16:48:47         # SparseLengthsMean
Jan 16 16:48:47         net.SparseLengthsMean(['dequantized_data', i, l],
Jan 16 16:48:47                               ['Mean_0'], engine='fp16')
Jan 16 16:48:47     
Jan 16 16:48:47         net.SparseLengthsMean8BitsRowwise(
Jan 16 16:48:47             ['quantized_data', i, l, 'scale_bias'],
Jan 16 16:48:47             ['Mean_1'])
Jan 16 16:48:47     
Jan 16 16:48:47         gathered_w = net.Gather(['quantized_data', i],
Jan 16 16:48:47                                 engine='fp16')
Jan 16 16:48:47     
Jan 16 16:48:47         gathered_scale_bias = net.Gather(['scale_bias', i],
Jan 16 16:48:47                                          engine='fp16')
Jan 16 16:48:47         net.Rowwise8BitQuantizedToFloat(
Jan 16 16:48:47             [gathered_w, gathered_scale_bias],
Jan 16 16:48:47             'Gathered_1')
Jan 16 16:48:47     
Jan 16 16:48:47         net.Gather(['dequantized_data', i], 'Gathered_0')
Jan 16 16:48:47     
Jan 16 16:48:47         workspace.GlobalInit(['caffe2', '--caffe2_log_level=0'])
Jan 16 16:48:47         workspace.RunNetOnce(init_net)
Jan 16 16:48:47         workspace.CreateNet(net)
Jan 16 16:48:47 >       workspace.RunNetOnce(net)
Jan 16 16:48:47 
Jan 16 16:48:47 ../.local/lib/python2.7/site-packages/caffe2/python/lengths_reducer_rowwise_8bit_ops_test.py:132: 
Jan 16 16:48:47 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Jan 16 16:48:47 ../.local/lib/python2.7/site-packages/caffe2/python/workspace.py:218: in RunNetOnce
Jan 16 16:48:47     StringifyProto(net),
Jan 16 16:48:47 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Jan 16 16:48:47 
Jan 16 16:48:47 func = <built-in method run_net_once of PyCapsule object at 0x7f94e450ca80>
Jan 16 16:48:47 op_id_fetcher = <bound method Workspace._last_failed_op_net_position of <caffe2.python.caffe2_pybind11_state_gpu.Workspace object at 0x7f952292cc00>>
Jan 16 16:48:47 net_name = 'bench_20'
Jan 16 16:48:47 args = ('\n\x08bench_20\x12P\n\x12init_6/UniformFill\x12\x0equantized_data\x12\nscale_bias\x1a\x00"\x1cFloatToRowwiseQuantize...init_6/Cast\x12\nGathered_0\x1a\x00"\x06Gather:\x12init_6/UniformFill:\x14init_6/UniformFill_2:\x0binit_6/Cast:\x01l',)
Jan 16 16:48:47 kwargs = {}, op_id = 8
Jan 16 16:48:47 net_tracebacks = {0: [('/var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/lengths_reducer_rowwise_8bit_ops_test.py', 84...py', 121, 'pytest_runtest_call'), ('/usr/lib/python2.7/site-packages/pluggy/callers.py', 187, '_multicall'), ...], ...}
Jan 16 16:48:47 tb = [('/var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/lengths_reducer_rowwise_8bit_ops_test.py', 119, '...unner.py', 121, 'pytest_runtest_call'), ('/usr/lib/python2.7/site-packages/pluggy/callers.py', 187, '_multicall'), ...]
Jan 16 16:48:47 line = ('/var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/lengths_reducer_rowwise_8bit_ops_test.py', 119, 'test_SparseSegmentUint8')
Jan 16 16:48:47 
Jan 16 16:48:47     def CallWithExceptionIntercept(func, op_id_fetcher, net_name, *args, **kwargs):
Jan 16 16:48:47         try:
Jan 16 16:48:47 >           return func(*args, **kwargs)
Jan 16 16:48:47 E           RuntimeError: [enforce fail at gather_op.h:195] . Gather is not implemented on tensor of type unsigned char Consider adding it a type in the list DispatchHelper or implementing a generic version (which won't work for duplicated indices though)
Jan 16 16:48:47 E           Error from operator: 
Jan 16 16:48:47 E           input: "quantized_data" input: "init_6/Cast" output: "bench_20/Gather" name: "" type: "Gather" engine: "fp16"frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7f94c08455c1 in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libc10.so)
Jan 16 16:48:47 E           frame #1: c10::ThrowEnforceNotMet(char const*, int, char const*, std::string const&, void const*) + 0x39 (0x7f94c08452e9 in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libc10.so)
Jan 16 16:48:47 E           frame #2: <unknown function> + 0x143dc3f (0x7f94e32a0c3f in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libcaffe2.so)
Jan 16 16:48:47 E           frame #3: <unknown function> + 0x143dd67 (0x7f94e32a0d67 in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libcaffe2.so)
Jan 16 16:48:47 E           frame #4: <unknown function> + 0x111c0fb (0x7f94e2f7f0fb in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libcaffe2.so)
Jan 16 16:48:47 E           frame #5: caffe2::SimpleNet::Run() + 0x165 (0x7f94e302d1d5 in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libcaffe2.so)
Jan 16 16:48:47 E           frame #6: caffe2::Workspace::RunNetOnce(caffe2::NetDef const&) + 0x2c (0x7f94e30623fc in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/../../torch/lib/libcaffe2.so)
Jan 16 16:48:47 E           frame #7: <unknown function> + 0x5109b (0x7f94e41b209b in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/caffe2_pybind11_state_gpu.so)
Jan 16 16:48:47 E           frame #8: <unknown function> + 0x8256c (0x7f94e41e356c in /var/lib/jenkins/.local/lib/python2.7/site-packages/caffe2/python/caffe2_pybind11_state_gpu.so)
Jan 16 16:48:47 E           <omitting python frames>
Jan 16 16:48:47 
Jan 16 16:48:47 ../.local/lib/python2.7/site-packages/caffe2/python/workspace.py:197: RuntimeError
```

**Comment by rjknight** [comment]
Looks like one of the CI runs didn't get updated with the status - "_GitHub rate limited us while adding a commit status of success: You.._"     Could someone retrigger the test? 

Thanks

**Comment by gchanan** [comment]
@rjknight unfortunately retest this please only works for jenkins tests.  I manually restarted the circleci test that wasn't updating the status.

**Comment by ezyang** [comment]
This is going to need tests, I think.

**Comment by ezyang** [comment]
Please remove wip tag when tests are added

**Comment by rjknight** [comment]
This seems like an unrelated failure
```ar 06 19:14:25 ======================================================================
Mar 06 19:14:25 FAIL: test_proper_exit (__main__.TestDataLoader)
Mar 06 19:14:25 There might be ConnectionResetError or leaked semaphore warning (due to dirty process exit), but they are all safe to ignore
Mar 06 19:14:25 ----------------------------------------------------------------------
Mar 06 19:14:25 Traceback (most recent call last):
Mar 06 19:14:25   File "/var/lib/jenkins/workspace/test/common_utils.py", line 120, in wrapper
Mar 06 19:14:25     fn(*args, **kwargs)
Mar 06 19:14:25   File "test_dataloader.py", line 774, in test_proper_exit
Mar 06 19:14:25     self.fail(fail_msg + ', and had no exception')
Mar 06 19:14:25 AssertionError: test_proper_exit with use_workers=True, pin_memory=False, hold_iter_reference=False, exit_method=worker_kill: loader process did not terminate, and had no exception
Mar 06 19:14:25 
Mar 06 19:14:25 ----------------------------------------------------------------------
Mar 06 19:14:25 Ran 51 tests in 31.381s
Mar 06 19:14:25 
Mar 06 19:14:25 FAILED (failures=1, skipped=10)
Mar 06 19:14:25 Traceback (most recent call last):
Mar 06 19:14:25   File "test/run_test.py", line 458, in <module>
Mar 06 19:14:25     main()
Mar 06 19:14:25   File "test/run_test.py", line 450, in main
Mar 06 19:14:25     raise RuntimeError(message)
Mar 06 19:14:25 RuntimeError: test_dataloader failed!
Mar 06 19:14:25 + cleanup
Mar 06 19:14:25 + retcode=1
Mar 06 19:14:25 + set +x
Mar 06 19:14:25 =================== sccache compilation log ===================
Mar 06 19:14:25 =========== If your build fails, please take a look at the log above for possible reasons ===========```

**Comment by rjknight** [comment]
CI fail looks like https://github.com/pytorch/pytorch/issues/16608

File: caffe2/operators/batch_gather_ops.cu [hunk 1]
```diff
@@ -16,8 +16,16 @@ bool BatchGatherOp<CUDAContext>::RunOnDevice() {
 template <>
 template <typename TInd>
 bool BatchGatherOp<CUDAContext>::DoRunWithType() {
+    return DispatchHelper<TensorTypes2<int8_t,int16_t,int32_t,int64_t,
+           uint8_t,at::Half,long,float,double,GenericTensorImplementation>,
```
Comment (gchanan) [inline caffe2/operators/batch_gather_ops.cu]: this has both `long` and `int64_t`?

Comment (rjknight) [reply caffe2/operators/batch_gather_ops.cu]: I added long to solve the original problem with the alexnet example, the original implementation allowed the types int32 and int64 as valid inputs, I think I just copy-pasted them - should I remove int64 and leave long?

**Pull request closed without merging**